### ISSUE TITLE
How to Convert format string `YYYY-MM-DD` to `std::chrono::time_point<std::chrono::system_clock, std::chrono::duration<int, std::ratio_multiply<std::ratio<24>, std::chrono::hours::period>>>`

### ISSUE THREAD

**Issue by zhengminlai** [HowardHinnant/date#689]
I have written a method to convert formatted string `YYYY-MM-DD HH-MM-SS` to `std::chrono::time_point<std::chrono::system_clock, std::chrono::microseconds>` like this:

```
std::chrono::time_point<std::chrono::system_clock, std::chrono::microseconds> StringToDateTime(const std::string &date) {
  std::istringstream in(date);
  ::date::sys_seconds tp;
  // format: 2000-06-04 14:11:00
  in >> ::date::parse("%F %T", tp);
  return tp;
}
```

But the follwing code fails to convert string `YYYY-MM-DD` to `std::chrono::time_point<std::chrono::system_clock, std::chrono::duration<int, std::ratio_multiply<std::ratio<24>, std::chrono::hours::period>>>`:

```
std::chrono::time_point<std::chrono::system_clock, std::chrono::duration<int, std::ratio_multiply<std::ratio<24>, std::chrono::hours::period>>> StringToDate(const std::string &date_str) {
  std::istringstream in(date_str);
  std::chrono::time_point<std::chrono::system_clock, std::chrono::duration<int, std::ratio_multiply<std::ratio<24>, std::chrono::hours::period>>>  dp;
  // format: 2000-06-04
  in >> date::parse("%F", dp);
  return dp;
}
```

Can you help me with that?  How to convert formatted string `YYYY-MM-DD` to `std::chrono::time_point<std::chrono::system_clock, std::chrono::duration<int, std::ratio_multiply<std::ratio<24>, std::chrono::hours::period>>>`? 

Thank you.

**Comment by HowardHinnant** [comment]
Your code works for me.  What error do you get?

**Comment by zhengminlai** [comment]
Uh, I have fixed it. Thank you. 

Another question, I am located in timezone CST +8. 

I write the following method to convert `std::chrono::time_point<std::chrono::system_clock, std::chrono::microseconds>`
to formatted string: 
```
std::string DateTimeToString(std::chrono::time_point<std::chrono::system_clock, std::chrono::microseconds>& date_time) {
  char buf[32];
  std::time_t now_c = std::chrono::system_clock::to_time_t(date_time);
  std::tm now_tm = *std::lcoaltime(&now_c);
  // format: 2000-06-04 14:11:00
  strftime(buf, 32, "%F %T", &now_tm);
  std::cout << std::string(buf));
  return std::string(buf);
}
```
The `date_time` parameter is fetched from the value in MySQL's `datetime` type. However, when the datetime string stored in MySQL is `2021-08-04 10:00:05`, I fetch that value and use the above method to convert `datetime` to string, the result string is `2021-08-04 18:00:05`.  And one more strange thing is that, when I use `std::gmtime(&now_c)`, it outputs the correct value. 

Do you know why `localtime` does not result in the correct time? 

Thanks.

**Comment by HowardHinnant** [comment]
MySQL is retrieving a local time.  The bug occurs when you store that local time in a `time_point<system_clock, microseconds>`.  This storage implicitly reinterprets the time as UTC.  Subsequently `std::localtime` adds your UTC offset to this value which outputs a value 8h too high.  `std::gmtime` converts to a UTC `tm`, doesn't add the UTC offset, and thus gets the right answer for the wrong reason.

To fix, either have MySQL retrieve UTC, and store that in your `system_clock`-based `time_point`, or store the MySQL in a `date::local_time<std::chrono::microseconds>`, and use `date::format` to format it.

_Aside:_  Before I get to that I thought I should mention that there exists some simplifying type-aliases which can make your code less verbose.

* `date::sys_time<std::chrono::microseconds>` is the same type as `std::chrono::time_point<std::chrono::system_clock, std::chrono::microseconds>`.
* `date::sys_days` is the same type as `std::chrono::time_point<std::chrono::system_clock, std::chrono::duration<int, std::ratio_multiply<std::ratio<24>, std::chrono::hours::period>>>`

Here is a rewritten `DateTimeToString` which takes a `date::local_time<std::chrono::microseconds>` and formats it:

```c++
std::string
DateTimeToString(date::local_time<std::chrono::microseconds> date_time)
{
    // format: 2000-06-04 14:11:00
    using namespace date;
    using namespace std::chrono;
    return format("%F %T", floor<seconds>(date_time));
}
```

You would also get the same result if you change the type of `date_time` back to `std::chrono::time_point<std::chrono::system_clock, std::chrono::microseconds>` (more concisely written as `date::sys_time<std::chrono::microseconds>`).  However like your current code it would be getting the right answer for the wrong reason.  By distinguishing local times from UTC times with the use of `local_time` vs `sys_time`, you harness the C++ type system to help you not confuse local times with UTC times.

In summary, `time_point`s based on `system_clock` **are** UTC.  These have a type alias `date::sys_time` for more concise code.  `date::local_time<some duration>` is used to hold local times.  My `tz.h` library can be used to convert between UTC and local times ([some installation is required](https://howardhinnant.github.io/date/tz.html#Installation)).

**Comment by zhengminlai** [comment]
Thank you so much for your help. 
 
I am using an `orm` framework called `sqlpp`, where it stores mysql's `datetime` type as `date::sys_time<std::chrono::microseconds>`. Therefore, it will take a lot of efforts to modify `sqlpp`'s source code to let it store the`datetime` type  as `date::local_time<std::chrono::microseconds>` (see https://github.com/rbock/sqlpp11/issues/374). In addition, in our company, we are not supposed to change Mysql's `time_zone`, and by default, Mysql's `time_zone` keeps consistent with `system_time_zone`, where in our region the `system_time_zone` is `CST`. That means we cannot let Mysql retrieve UTC time in our project.

Hence, it seems that we can only fix the time conversion by using `gmtime` to get the right answer for the wrong reason?

**Comment by HowardHinnant** [comment]
`gmtime` or:
```c++
std::string
DateTimeToString(date::sys_time<std::chrono::microseconds> date_time)
{
    // format: 2000-06-04 14:11:00
    using namespace date;
    using namespace std::chrono;
    return format("%F %T", floor<seconds>(date_time));
}
```
(note I've changed the argument to `sys_time`).

They are both equivalent, converting a count of seconds into a human readable UTC timestamp, with no regard for time zones.

**Comment by zhengminlai** [comment]
Thank you so much.
